### DISCUSSION TITLE
FilamentPHP 3 Table | Check-In Check-Out in one row?

### DISCUSSION THREAD

**Discussion by madushancs** [filamentphp/filament#8981] [Help] [answered]
I have a MySQL data table called attendance. column names are id, employee_id, date and time.

if employee_id 1 check-in at 8.00 AM and check-out at 5.00 PM on the same day. So I have two records as below.

```
ID  | employee_id    | date             | time
1   |      1         | 2023-10-09       | 08:00:00
2   |      1         | 2023-10-09       | 17:00:00
```

If I want to show check-in & check-out in the same row in the FilamentPHP table, how can I do it?

**Comment by leandrocfe** [comment]
[formatStateUsing](https://filamentphp.com/docs/3.x/tables/columns/text#custom-formatting) ?

**Reply by madushancs** [reply]
I tied but couldn't. Any advise?

**Comment by joshembling** [comment]
Try chaining this 

```php
->getStateUsing(function (Model $record) {
    // return whatever you need to show e.g.
    return "In: {$record->checked_in} Out: {$record->checked_out};
})
```

**Reply by madushancs** [reply]
Hi Joshembiling, Actually I want to show extra columns not like the attached image. 
![image](https://github.com/filamentphp/filament/assets/45415384/689546c3-c369-4f6f-80bd-705d65e0f54b)

I have explained more information in here https://github.com/filamentphp/filament/discussions/8981#discussioncomment-7241765
Do you have any idea about this?

**Comment by leandrocfe** [comment]
hum, would you like to show 
`employee_id    |     check_in      |      check_out`
grouped by employee_id, right?

Would it be a readonly table?

**Reply by madushancs** [reply]
Hi Leandrocfe, There is a fingerprint machine. every day employees come in the morning and go out in the evening (8.00 AM to 5.00 PM). each **in** and **out** they use the fingerprint machine and save each record in the attendance table. attendance table only have **ID**, **employee_id**, **date** and **time**. column. there are **no** check-in & check-out columns. So I want groupBy date employee_id and show check-in check-out in the same row using record. 

for example, below have some test records
```
attendance Table (MySQL)
ID  | employee_id    | date             | time
1   |      1         | 2023-10-09       | 08:00:00
2   |      1         | 2023-10-09       | 17:00:00
3   |      2         | 2023-10-09       | 07:50:00
4   |      2         | 2023-10-09       | 17:10:00
```

Using the above data I want to show below table
```
FilamentPHP table (View)
Date               Employee ID     Check-in      Check-out
2023-10-09         1               08:00:00      17:00:00
2023-10-09         2               07:50:00      17:10:00
```

**Reply by jinbatsu** [reply]
It can be grouped by two column: empyee_id and date.
And select using MIN() and MAX() to determine as ceck-in and check-out.

**Reply by madushancs** [reply]
could you please help me do it with filament 3?  

do you mean like this
```
    public static function query($query)
    {
        return $query->groupBy(['date', 'employeeid'])
            ->select([
                'date',
                'employeeid',
                DB::raw('MIN(time) as check_in_time'),
                DB::raw('MAX(time) as check_out_time'),
            ]);
    }
```

**Reply by leandrocfe** [reply]
Could you try this please?

```php
public static function table(Table $table): Table
{
    return $table
        ->defaultGroup('employeeid')
        ->groupsOnly()
        ->columns([
            TextColumn::make('checkin')
                ->summarize(
                    Summarizer::make()
                        ->using(fn (\Illuminate\Database\Query\Builder $query) => $query->min('time'))
                ),
            TextColumn::make('checkout')
                ->summarize(
                    Summarizer::make()
                        ->using(fn (\Illuminate\Database\Query\Builder $query) => $query->max('time'))
                ),
        ]);
}
```

[reference](https://filamentphp.com/docs/3.x/tables/summaries)

**Reply by madushancs** [reply]
Hi leandrocfe, Thanks for supporting. here what I got. 
![image](https://github.com/filamentphp/filament/assets/45415384/92c14c98-4309-44c7-b2ef-a4ccd39e5200)
can we group by date also same time because if the same employee_id come on two different date it not showing here?

**Reply by leandrocfe** [reply]
Did you try to add this column?

**Reply by madushancs** [reply]
yes just copy and paste your code ☺️

**Reply by leandrocfe** [reply]
I mean, `TextColumn::make('date')`?

**Reply by madushancs** [reply]
Yes, here is the result. 
![image](https://github.com/filamentphp/filament/assets/45415384/7756c6c3-765c-4099-997c-198fa8c49b5c)

**Reply by madushancs** [reply]
can we do something like that,

public static function query($query)
    {
        return $query->groupBy(['date', 'employeeid'])
            ->select([
                'date',
                'employeeid',
                DB::raw('MIN(time) as check_in_time'),
                DB::raw('MAX(time) as check_out_time'),
            ]);
    }

and that query passes to the table method. I don't know how to do that. can you help me?

**Reply by leandrocfe** [reply]
Not sure if you can use [modifyQueryUsing](https://filamentphp.com/docs/3.x/panels/resources/listing-records#customizing-the-table-eloquent-query)
Could you share the project on github?

**Reply by madushancs** [reply]
I have sent an invite.  could you please check it? normally attendance records come from a fingerprint machine. For test purposes can add manually. Thanks

**Reply by leandrocfe** [reply, marked as answer]
A workaround

AttendanceResource.php
```php
public static function table(Table $table): Table
{
    return $table
        ->columns([
            TextColumn::make('id')->label('Employee ID'),
            TextColumn::make('date'),
            TextColumn::make('check_in_time'),
            TextColumn::make('check_out_time'),
        ]);
}

public static function getEloquentQuery(): Builder
{
    return Attendance::query()
        ->selectRaw('
            employeeid as id,
            date,
            min(time) as check_in_time,
            max(time) as check_out_time
        ')
        ->groupByRaw('employeeid, date');
}
```

Pages/ListAttendances.php
```php
use Illuminate\Database\Eloquent\Builder;
protected function applySortingToTableQuery(Builder $query): Builder
{
    return $query;
}
```

**Reply by madushancs** [reply]
Hi leandrocfe, That works. Thank you very much for your time and great support. really appreciate it.
